archival: Fix archival_stm_snapshot installation

Lazin opened the pull request:
The code that installs the archival STM snapshot during the recovery uses the wrong offset to seed the in-sync offset of the STM (off by one). Because of that the STM may skip the first record batch (which is usually a raft_configuration batch which doesn't affect the archival STM).

Also, the last applied offset of the snapshot is seeded incorrectly using the last applied offset of the manifest. The commit fixes this by using last uploaded offset.

Finally, the commit fixes the ducktape test by adding a short sleep between the deletion of the topic and recovery of the topic. When this two operations are performed back to back it's possible for the recover to pick up different version of the manifest on different replicas. One of the replicas might be uploading the manifest concurrently while the topic is deleted and created which will lead to divergence. This is not a problem during the actual disaster recovery.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.1.x
- [x] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
* none

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/64697-->
<!--2a33da29-7a4c-47cb-bda3-7460ff9a17aa-->
<!--title-->
## Retry command for [Build#64697](https://buildkite.com/redpanda/redpanda/builds/64697)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_migrated_topic_data_integrity@{"params":{"cancellation":{"dir":"in","stage":"preparing"},"use_alias":true},"transfer_leadership":true}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--66113657-131a-4779-bf7e-7ab54bdf74de-->
<!--TEST RESULTS-->
<!--80aaf7ff-9a2b-479a-9a1c-8f555215db29-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64697">build#64697</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.True | ducktape | https://buildkite.com/redpanda/redpanda/builds/64697#01964540-8ae5-4578-b387-944f975875c7 | FLAKY | 16/21 |
| test_archival_service_rpfixture.test_archival_service_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/64697#019644fe-c397-45fc-9bce-23416fae44ef | FAIL | 0/2 |
| test_archival_service_rpfixture.test_archival_service_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/64697#019644fe-c398-4b99-a765-d0cee91a3c84 | FAIL | 0/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64755">build#64755</a></summary>


| test_id                                                                                                                                                                             | test_kind   | job_url                                                                                   | test_status   | passed   |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| gtest_raft_rpunit.gtest_raft_rpunit                                                                                                                                                 | unit        | https://buildkite.com/redpanda/redpanda/builds/64755#01964902-fc17-4d87-87b3-f411b0f2e056 | FLAKY         | 1/2      |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.False | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64755#01964944-0a0f-4b37-9fdf-33e1db953ea4 | FLAKY         | 17/21    |
| rptest.tests.datalake.disk_budget_test.DatalakeDiskUsageTest.test_idle_finish.num_partitions=30.concurrent_translations=1.cloud_storage_type=CloudStorageType.S3                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64755#01964947-a492-41ae-b05c-3671ac864673 | FLAKY         | 8/21     |
| rptest.tests.datalake.throttling_test.DatalakeThrottlingTest.test_basic_throttling.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64755#01964947-a492-4f15-af56-0c125a3ff82c | FLAKY         | 18/21    |
| rptest.tests.datalake.throttling_test.DatalakeThrottlingTest.test_basic_throttling.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP                      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64755#01964947-a492-41ae-b05c-3671ac864673 | FLAKY         | 19/21    |
</details>
<!--/body-->

mmaslankaprv commented on tests/rptest/tests/scaling_up_test.py:
```diff
@@ -670,6 +670,8 @@ def test_scaling_up_with_recovered_topic(self):
         total_replicas = 3 * partition_count
         rpk.delete_topic(topic.name)
 
+        time.sleep(10)
```
nit: can this be a separate commit ?

Lazin replied:
done

Lazin committed:
archival: Fix archival STM snapshot installation

The code that installs the archival STM snapshot during the recovery
uses the wrong offset to seed the in-sync offset of the STM (off by
one). Because of that the STM may skip the first record batch (which is
usually a raft_configuration batch which doesn't affect the archival
STM).

Also, the last applied offset of the snapshot is seeded incorrectly
using the last applied offset of the manifest. The commit fixes this by
using last uploaded offset.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
rptest: Fix scaling_up_test.py

This commit fixes the ducktape test by adding a short sleep
between the deletion of the topic and recovery of the topic. When this
two operations are performed back to back it's possible for the recover
to pick up different version of the manifest on different replicas. One
of the replicas might be uploading the manifest concurrently while the
topic is deleted and created which will lead to divergence. This is not
a problem during the actual disaster recovery.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

mmaslankaprv commented:
seems that there is regression in unit tests

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/64755-->
<!--2d88c93a-99c4-47e2-bbf4-f511a7d9c1f0-->
<!--4502357d-5758-42e9-8676-c9aa77364f61-->
<!--title-->
## Retry command for [Build#64755](https://buildkite.com/redpanda/redpanda/builds/64755)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```

/ci-repeat 1
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_migrated_topic_data_integrity@{"params":{"cancellation":{"dir":"in","stage":"preparing"},"use_alias":false},"transfer_leadership":true}
tests/rptest/tests/datalake/disk_budget_test.py::DatalakeDiskUsageTest.test_idle_finish@{"cloud_storage_type":1,"concurrent_translations":1,"num_partitions":30}
```
<!--/body-->

Lazin commented:
/ci-repeat 1
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_migrated_topic_data_integrity@{"params":{"cancellation":{"dir":"in","stage":"preparing"},"use_alias":false},"transfer_leadership":true}
tests/rptest/tests/datalake/disk_budget_test.py::DatalakeDiskUsageTest.test_idle_finish@{"cloud_storage_type":1,"concurrent_translations":1,"num_partitions":30}

Lazin commented:
/ci-repeat 1
tests/rptest/tests/datalake/throttling_test.py

vbotbuildovich commented:
/backport v25.1.x

vbotbuildovich commented:
/backport v24.3.x
